### PR TITLE
FS-728 - Add round data

### PR DESCRIPTION
**Pull request by harryyo**
…data to applications. assessment_deadline updated to resolved from round store

### CODE REVIEW

File: api/namespace/application/application_ns.py [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
 
-
 """
 Namespace acts as part of an api (with the same methods as api)
 """
 
 
-search_ns = Namespace("search", description="application search operations")
\ No newline at end of file
+application_ns = Namespace("application", path="", description="single application operations")
```
Comment (SentientHousePlant) [inline api/namespace/application/application_ns.py:8]: Nice that there are separate endpoints for single and bulk operations.

File: api/namespace/application/routes.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+@application_ns.route("/<application_id>", methods=["GET", "POST"])
+class Application(Resource):
+    """
+    Operations on a single application
+    """
+
+    @application_ns.doc("get_application")
+    @application_ns.marshal_with(application_full, code=200)
+    def get(self, application_id):
+        return APPLICATIONS.get_application(application_id), 200
```
Comment (SentientHousePlant) [inline api/namespace/application/routes.py:33]: Nice abstractions as always.

File: api/namespace/applications/applications_ns.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from flask_restx import Namespace
+
+
+"""
```
Comment (SentientHousePlant) [inline api/namespace/applications/applications_ns.py:4]: Should be above imports

File: config.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+"""
+Application Config
+"""
+SECRET_KEY = environ.get("SECRET_KEY") or "dev"
+SESSION_COOKIE_NAME = environ.get("SESSION_COOKIE_NAME") or "session_cookie"
+FLASK_ROOT = path.dirname(path.realpath(__file__))
+FLASK_ENV = environ.get("FLASK_ENV") or "development"
+
+"""
```
Comment (SentientHousePlant) [inline config.py:13]: Not a fan of using docstrings instead of standard comments.

File: database/store.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                    "question": question.get("question"),
+                    "status": question.get("status"),
+                })
+            application_summary.update({"questions": questions})
+            return application_summary
+
+    def update_status(
+        self, application_id: str, question_name: str, new_status: str
+    ) -> dict or False:
+        """
```
Comment (SentientHousePlant) [inline database/store.py:96]: Great docstrings.

File: database/store.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        datetime_end = params.get("datetime_end")
+        fund_id = params.get("fund_id")
+        status_only = params.get("status_only")
+        id_contains = params.get("id_contains")
+        order_by = params.get("order_by", "id")
+        order_rev = params.get("order_rev") or False
+        try:
+            order_rev = strtobool(order_rev)
+        except ValueError:
+            pass
```
Comment (SentientHousePlant) [inline database/store.py:128]: 👀 any way to bypass this try except block - sorry these just give me the creeps.

File: external_data/data.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    """
+    fund_round = None
+    if round_id:
+        round_endpoint = ROUND_STORE_API_HOST + ROUND_ENDPOINT.format(
+            fund_id=fund_id, round_id=round_id
+        )
+        round_response = get_data(round_endpoint)
+        if round_response and "round_id" in round_response:
+            fund_round = Round.from_json(round_response)
+        if not fund_round:
```
Comment (SentientHousePlant) [inline external_data/data.py:94]: Ok, lets have this debate again...I don't like that it's not explicit that a comparison to None is being done here. If you disagree that's fine but I would prefer 
```python
if fund_round is not None:
```

File: external_data/models/fund.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
+from dataclasses import dataclass
+from typing import List
+
+from external_data.models.round import Round
+
+
+@dataclass
```
Comment (SentientHousePlant) [inline external_data/models/fund.py:7]: Lovely that you're using dataclass....check out attrs - its already included in our requirements by other libraries and _some_ people think it's better then dataclass. It's just slightly cooler and has more batteries included.

File: external_data/models/round.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    title: str
+    identifier: str
+    fund_id: str
+    opens: datetime
+    deadline: datetime
+    assessment_deadline: datetime
+    eligibility_criteria: dict
+
+    @staticmethod
+    def from_json(data: dict):
```
Comment (SentientHousePlant) [inline external_data/models/round.py:16]: Nice factory

File: tests/helpers.py [hunk 10]
```diff
@@ -16,9 +18,23 @@ def expected_data_within_get_response(
     """
     response = test_client.get(endpoint, follow_redirects=True)
     response_data = json.loads(response.data)
-    print(endpoint)
-    print(response_data)
-    assert response_data == expected_data, response_data
+
+    if isinstance(expected_data, List):
```
Comment (SentientHousePlant) [inline tests/helpers.py:22]: Ok, I've got to be honest - this is quite hard to read and understand. Any chance of adding some comments explaining what is happening? Or perhaps I could challenge you to write this in a more pythonic and elegant way?

**Review by SentientHousePlant — CHANGES_REQUESTED** [review]
Very nice abstractions and use of the standard library. 

Some small criticism towards the end of the review regarding some test helpers. 

Overall very good and well structured.

File: api/namespace/application/application_ns.py (hunk 1 above)
Comment (Adam-W1) [reply api/namespace/application/application_ns.py:8]: agreed - much nicer to call. great work Harry

File: api/namespace/application/routes.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+
+@application_ns.route("/", methods=["GET", "POST"])
+class ApplicationCreate(Resource):
+    """
+    Create a new application
+    """
+
+    @application_ns.doc("create_application")
+    @application_ns.expect(application_inbound)
```
Comment (Adam-W1) [inline api/namespace/application/routes.py:18]: we may want to pass in a boolean to @expect here to verify the payload (according to RESTX docs: https://flask-restplus.readthedocs.io/en/stable/swagger.html#the-api-expect-decorator)

File: database/store.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+from dateutil.tz import UTC
+from slugify import slugify
+from operator import itemgetter
+from distutils.util import strtobool
+from external_data.data import get_round
+
+
+class ApplicationDataAccessObject(object):
+    """
+    return all funds (replace the date object with a string to send)
```
Comment (Adam-W1) [inline database/store.py:21]: this is just me I'm sure but i dont get what this is saying?

File: database/store.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+    """
+
+    def __init__(self):
+        self.applications: dict = initial_application_store_state
```
Comment (Adam-W1) [inline database/store.py:25]: another newbie python Q, should this be a private var, i.e: self._applications ... ? or maybe have a @property decorator. Please tell me if this is stupid

File: api/namespace/application/routes.py (hunk 2 above)
Comment (harryyo) [reply api/namespace/application/routes.py:33]: Thanks

File: external_data/models/round.py (hunk 9 above)
Comment (harryyo) [reply external_data/models/round.py:16]: Thanks

File: database/store.py (hunk 6 above)
Comment (harryyo) [reply database/store.py:128]: `hasattr(params, "order_rev")` ?

Comment (harryyo) [reply database/store.py:128]: `params.get("order_rev") == "1"` may be better

File: external_data/data.py (hunk 7 above)
Comment (harryyo) [reply external_data/data.py:94]: Well I'd argue that two wrongs don't make a right

File: tests/helpers.py (hunk 10 above)
Comment (harryyo) [reply tests/helpers.py:22]: Yep is horrific! I need a nested solution to go into two dictionaries and iterate through items and check matches for both dicts and lists - will work on

File: .idea/vcs.xml [hunk 14]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="VcsDirectoryMappings">
+    <mapping directory="$PROJECT_DIR$" vcs="Git" />
+  </component>
+</project>
```
[the remark this answers is not in the document]
Comment (harryyo) [reply .idea/vcs.xml]: The **one** time I do git add . and there's no .idea in the gitignore. Apols! Removed obv - gitignore updated for intellij fans everywhere

File: tests/helpers.py (hunk 10 above)
Comment (harryyo) [reply tests/helpers.py:22]: Not exactly rising to the challenge - but found a package DeepDiff that does deep dictionary comparison, which should be more manageable

File: api/namespace/application/models/application.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+application_full = application_ns.model(
+    "application_full",
+    {
+        "id": fields.String(
+            description="The id of the application.",
+        ),
+        "status": fields.String(
+            description="The status of the application",
+            example="NOT_STARTED"
```
[the remark this answers is not in the document]
Comment (harryyo) [reply api/namespace/application/models/application.py:41]: Total slapdashery - thanks for spotting. Updated

File: database/store.py (hunk 12 above)
Comment (harryyo) [reply database/store.py:21]: I think this is an old dev comment left in and moved here in error - removed

File: external_data/models/fund.py (hunk 8 above)
Comment (harryyo) [reply external_data/models/fund.py:7]: Noted. Maybe we need some use cases where our classes are doing more hard work, validation etc so we can get out the big guns?

File: database/store.py (hunk 13 above)
Comment (harryyo) [reply database/store.py:25]: I think good suggestion. Privatised.

File: api/namespace/application/models/application.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+        ),
+        "fund_id": fields.String(
+            description="The id of the fund",
+            example="funding-service-design"
+        ),
+        "fund_name": fields.String(
```
Comment (stevewalton28) [inline api/namespace/application/models/application.py:47]: Do we need this? we have the fund_id - if the fund name changes we have to update it here as well - the fund-store should be our source of truth for all information about the fund

File: api/namespace/application/models/application.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
```
Comment (stevewalton28) [inline api/namespace/application/models/application.py:92]: Same as above - should only hold the fund-id

File: api/namespace/applications/models/applications.py [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+        ),
+        "status": fields.String(
+            description="The status of the application",
+            example="NOT_STARTED"
+        ),
+        "fund_id": fields.String(
+            description="The id of the fund",
+            example="funding-service-design"
+        ),
+        "fund_name": fields.String(
```
Comment (stevewalton28) [inline api/namespace/applications/models/applications.py:18]: Shouldn't hold fund name see abovce

File: database/store.py [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+
+    @property
+    def applications_index(self) -> dict:
+        applications = {}
+        for application_id, application in self._applications.items():
+            application_summary = {
+                "id": application.get("id"),
+                "status": application.get("status"),
+                "fund_id": application.get("fund_id"),
+                "fund_name": application.get("fund_name"),
```
Comment (stevewalton28) [inline database/store.py:28]: Shouldn't hold fund name here

**Review by stevewalton28 — CHANGES_REQUESTED** [review]

File: api/namespace/application/models/application.py (hunk 16 above)
Comment (harryyo) [reply api/namespace/application/models/application.py:47]: Ok, appreciate that single source of truth, however when we're pulling a list of applications from multiple funds (as the assessment frontend does), if we wanted fund_name for each application we would have to make a separate api call for each row (or cache fund_store data on assessment frontend).

Comment (harryyo) [reply api/namespace/application/models/application.py:47]: Is there a preferred way for collating info like this for related objects when returning a list?

Comment (stevewalton28) [reply api/namespace/application/models/application.py:47]: Seperate API calls are expected!

**Review by SentientHousePlant — APPROVED** [review]
LGTM!!

**Review by ramsharma-prog — APPROVED** [review]

**Pull request merged**